Node revamp: PostgreSQL lessons (#27960)
TheOdinProject/curriculum: nodeJS/express/using_postgresql.md

<!-- markdownlint-disable TOP004 -->

### Introduction

Now that we have PostgreSQl up and running on our machine, it's time to work with it.

<div class="lesson-note lesson-note--critical" markdown="1">

#### Prerequisite SQL knowledge

Make sure you've completed the [SQL course](https://www.theodinproject.com/paths/full-stack-javascript/courses/databases).

This lesson and all subsequent lessons will assume you understand SQL syntax and concepts.

</div>

For brevity's sake, we'll refer to database as db from now on.

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

- Setting up a new db in PostgreSQL shell
- Setting up and querying with node-postgres
- Using a script to populate the db

### Setting up a Express app

Let's start by creating a Express application. It will just have one feature - add usernames provided by the user to the db. Here are the expected routes and their functionalities:

> - `GET /` - to log available usernames in the DB to the terminal. For now, just put in a simple `console.log('usernames will be logged here - wip')`.

Reviewer: For consistency since we typically use double quotes in our JS examples and this is the sole use of JS single quotes in this lesson :P

```suggestion
- `GET /` - to log available usernames in the DB to the terminal. For now, just put in a simple `console.log("usernames will be logged here - wip")`.
```

> - `GET /new` - to display a HTML form to the user with one username input text field. It will submit to the following route,

Reviewer: Might look a tiny bit cleaner than a trailing comma?

```suggestion
- `GET /new` - to display a HTML form to the user with one username input text field. It will submit to the next route.
```

- `POST /new` - will save the incoming username data to the DB. For now, just log `console.log("username to be saved: ", req.body.username)`.

> Move on to the next section once you have the above functionality working. Scope relevant code to the `routes` and `controllers` folders. Since there's only one view (`GET /new`) to take care of, it's up to you whether you want to use `ejs` or make do with plain ol' HTML.

Reviewer: Here and line 106, should we change folder/folders to directory/directories?

Author: is there a particular reason to switch?

Reviewer: My understanding was that "directory" is more technically correct as "folder" technically refers to the GUI representation of a directory. But having said that, it looks like we use "folder" a lot throughout the curriculum anyway, so no biggie at all.

```sql
CREATE TABLE usernames (
   id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
   username VARCHAR ( 255 ) 
);
```

Verify that the table has been created by running `\d`. You should see the following two tables in the output (we've skipped some output details for brevity):

```bash
|       Name       |   Type   |
+ -----------------+----------+
| usernames        | table    |
| usernames_id_seq | sequence |
```

<div class="lesson-note" markdown="1">

#### Identity column

Wait a minute, what's this `usernames_id_seq` thing?

The `GENERATED ALWAYS AS IDENTITY` clause is the culprit. It defined the `id` column as an [identity column](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-PARMS-GENERATED-IDENTITY). PostgreSQL now automatically generates a value for this column. By default it starts at 1 and increments by 1 for each new row. Additionally, PostgreSQL implicitly creates `usernames_id_seq`, which is a sequence object, that keeps track of the next value to be used.

</div>

Woohoo, we now have a db and a table... a lonely table. Not for long:

```sql
INSERT INTO usernames (username)

> VALUES ('Mao'), ('nevz'), ('Lofty');

Reviewer: :heart:

```

Verify:

```sql
SELECT * from usernames;
```

### Using node-postgres in Express

> We can work with PostgreSQL in our Express application through [`node-postgres`](https://github.com/brianc/node-postgres) (or `pg` for short). It is library that we'll use to interface with the PostgreSQL db. Install it by:

Reviewer: ```suggestion
We can work with PostgreSQL in our Express application through [`node-postgres`](https://github.com/brianc/node-postgres) (or `pg` for short). It is library that we'll use to interface with the PostgreSQL db. Install it with:
```
Or "Install it by running:"

module.exports = new Pool({
  connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users"
});
```

Pick whichever method you want to use and let's continue.

<div class="lesson-note" markdown="1">

#### Two ways of connecting with pg

`pg` has two ways to connect to a db: a client and a pool.

Client is an individual connection to the DB, which you manually manage. You open a connection, do your query, then close it. This is fine for one-off queries, but can become expensive if you're dealing with a lot of queries. Wouldn't this problem be alleviated if we could somehow hold onto a client? Yes!

Enter pool. As the name suggests, it's a pool of clients. A pool holds onto connections. And when you query, it'll programmatically open a new connection unless there's an existing spare one. Perfect for web servers.

</div>

### Querying with pg

With our initialized `Pool`, we can use the `query` method. Create a new `db/queries.js` file. Upon revising our project requirements, we understand we need two db interactions: getting all usernames and inserting a new username. Let's define these functions:

```javascript

> const pool = require("./pool");

Author: I intended for this to stay as `pool`, but ended up using `db` in the code examples. Let me update to `pool`.

const pool = require("./pool");

async function getAllUsernames() {
  const { rows } = await db.query("SELECT * FROM usernames");
  return rows;
}

async function insertUsername(username) {
  await db.query("INSERT INTO usernames (username) VALUES ($1)", [username]);
}

module.exports = {
  getAllUsernames,
  insertUsername

> }

Reviewer: Nit for inter-lesson consistency
```suggestion
};
```

```

<div class="lesson-note" markdown="1">

#### Parameterization

What's with the `$1` in the insert query?

Alternatively, the query could look like:

```javascript
await db.query("INSERT INTO usernames (username) VALUES ('" + username + "')");
```

We're passing user entered value i.e. `username` directly into our query. A nefarious user could enter something like `sike'); DROP TABLE usernames; --` and wreak havoc. Scary stuff. This is called [SQL injection](https://en.wikipedia.org/wiki/SQL_injection).

`pg` provides [query parameterization](https://node-postgres.com/features/queries#parameterized-query) to prevent this. Instead of passing user input directly, we pass it in an array as the second argument. `pg` handles the rest.

</div>

> Invoke the above two functions in the specific controllers (you might have different function names etc. Important thing is to understand how the db functions are invoked):

Reviewer: ```suggestion
Invoke the above two functions in the specific controllers (you might have different function names etc. The important thing is to understand how the db functions are invoked):
```

```javascript

> const db = require("../db/queries");

Reviewer 2: I've commented above that named the pool module as `db` should we name this as `query` instead

```javascript
await query.getAllUsernames();

await query.insertUsername(username);
```
Or we can make it more modular by creating `db/queries/user.js` and have generic names like `Username.getAll()` `Username.insert()` but it's up to you

What do you think?

Author: I'd rather keep it as `db`, though at the end of the day I really don't feel strongly towards it. I've updated the other code example., doo let me know what you prefer though after you're read that one; I'll update the lesson accordingly.

Regarding your suggestion on making it modular: I'll trust the learners to refactor as they see fit when they work on bigger projects. They already have all the knowledge they need to do `Username.getAll()` or however else they want to refactor.

Author: ^ adding to the above, you might've also noticed how barebones the script was compared to the version you wrote (which included utility functions etc). I'd rather have learners come up with this stuff

> ```javascript
> const db = require("../db/queries");
> 
> async function getUsernames(req, res) {
>   const usernames = await db.getAllUsernames();
>   console.log("Usernames: ", usernames);
>   res.send("Usernames: " + usernames.map(user => user.username).join(", "));
> }
> 
> async function createUsernameGet(req, res) {
>   // render the form
> }
> 
> async function createUsernamePost(req, res) {
>   const { username } = req.body;
>   await db.insertUsername(username);
>   res.redirect("/");
> }
> 
> module.exports {
>   getUsernames,
>   createUsername
> }
> ```

Reviewer: Since these are middleware that can throw, we should use `express-async-handler` as is introduced in the [Controllers lesson](https://github.com/TheOdinProject/curriculum/pull/27695).

Also fixing the exports bit

````suggestion
```javascript
const db = require("../db/queries");
const asyncHandler = require("express-async-handler");

const getUsernames = asyncHandler(async (req, res) => {
  const usernames = await db.getAllUsernames();
  console.log("Usernames: ", usernames);
  res.send("Usernames: " + usernames.map(user => user.username).join(", "));
});

const createUsernameGet = asyncHandler(async (req, res) => {
  // render the form
});

const createUsernamePost = asyncHandler(async (req, res) => {
  const { username } = req.body;
  await db.insertUsername(username);
  res.redirect("/");
});

module.exports = {
  getUsernames,
  createUsernameGet,
  createUsernamePost,
};
```
````

Author: I've skipped A LOT of things by design like error handling, data validation, and `express-async-handler`. This code example is just there to show how the learner might use their database query functions, that's it. The complete controller implementation is on them.

Reviewer: That's fair enough then. We'd still need to fix the module.exports bit for the `=` and right function names

```js
module.exports = {
  getUsernames,
  createUsernameGet,
  createUsernamePost,
};
```

Author: Adding to the above, I'm not expecting learners to implement everything they've learned in the controllers lesson for the app they build this lesson. The purpose for this lesson's app is to practice db stuff only.

Author: fixed

Take your app for a spin, hopefully it works as expected.

### Populate the db via a script

> You might have noticed how cumbersome it is to create a table and populate it with data. Luckily, we have the power of c(n)ode by our side, let's automate it via a script. Create a new file `db/populatedb.js`.

Reviewer: Should we also explicitly say here about replacing the connection string for the Client as necessary?

Side note: `c(n)ode` lmao :joy:

Author: > Should we also explicitly say here about replacing the connection string for the Client as necessary?

^ not sure what you mean

Reviewer: Referring to the Client connection string in the example below
```js
  const client = new Client({
    connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users",
  });
  ```
Not a big thing really, only wondering if it wouldn't hurt to add a little mention either here or as a comment in the example like 
`// replace with your own connection string`
or something

Author: they shall experience the pain of blindly copy pasting stuff 🙈

  });
  await client.connect();
  await client.query(SQL);
  await client.end();
  console.log("done");
}

main();
```

Login to the PostgreSQL shell, connect to the `top_users` db, and drop the `usernames` table:

```sql
DROP TABLE usernames;
```

You can then run this script via `node db/populatedb.js`, or add it as a [`script` in `package.json`](https://stackoverflow.com/a/36433748).

Do note that the script is designed to be ran only once.

#### Populating production dbs

We've hardcoded our local db connection information in the script. Hence, the script only populates our local db. We need a way to populate our production db as well. One way to this is by using environment variables, though this leads to unnecessary hassle. Why? Because now, the script can only populate the production db on the production server i.e. we'll need to access the production server's cli to run the script. Or, we could sneakily edit our environment file to point to the production db and run the script on our machine, and revert it.

We should aim to make our script as independent from our codebase as possible.

> A far painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [`process.argv`](https://nodejs.org/docs/latest/api/process.html#processargv).

Reviewer: **Love** the `process.argv` introduction here. Can't believe the current course doesn't mention it unless you encounter it by chance and remember it through the early Node docs.

Grammar nit

```suggestion
A far more painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [`process.argv`](https://nodejs.org/docs/latest/api/process.html#processargv).
```

> You can then run this script via `node db/populatedb.js`, or add it as a [`script` in `package.json`](https://stackoverflow.com/a/36433748).
> 
> Do note that the script is designed to be ran only once.
> 
> #### Populating production dbs
> 
> We've hardcoded our local db connection information in the script. Hence, the script only populates our local db. We need a way to populate our production db as well. One way to this is by using environment variables, though this leads to unnecessary hassle. Why? Because now, the script can only populate the production db on the production server i.e. we'll need to access the production server's cli to run the script. Or, we could sneakily edit our environment file to point to the production db and run the script on our machine, and revert it.
> 
> We should aim to make our script as independent from our codebase as possible.
> 
> A far more painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [`process.argv`](https://nodejs.org/docs/latest/api/process.html#processargv).

Reviewer 2: Nit:
```suggestion
You can then run this script via `node db/populatedb.js`, or add it as a [script in package.json](https://stackoverflow.com/a/36433748).

Do note that the script is designed to be ran only once.

#### Populating production dbs

We've hardcoded our local db connection information in the script. Hence, the script only populates our local db. We need a way to populate our production db as well. One way to this is by using environment variables, though this leads to unnecessary hassle. Why? Because now, the script can only populate the production db on the production server i.e. we'll need to access the production server's cli to run the script. Or, we could sneakily edit our environment file to point to the production db and run the script on our machine, and revert it.

We should aim to make our script as independent from our codebase as possible.

A far more painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [process.argv](https://nodejs.org/docs/latest/api/process.html#processargv).
```
Do you think it's better to not use backticks for these two?

![image](https://github.com/TheOdinProject/curriculum/assets/75290989/0a5d8adf-9eca-471d-aa0b-a159f87f5ac3)

![image](https://github.com/TheOdinProject/curriculum/assets/75290989/d324079b-cde1-4e58-b961-084e9e021fcc)

> ```bash
> # populating local db 
> node db/populatedb.js local-db-url
> 
> # populating production db
> # run it from your machine once after deployment of your app & db
> node db/populatedb.js production-db-url
> ```

Reviewer: Is this what's intended?

````suggestion
```bash
# populating local db 
node db/populatedb.js <local-db-url>

# populating production db
# run it from your machine once after deployment of your app & db
node db/populatedb.js <production-db-url>
```
````

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Skim through [pg's documentation](https://node-postgres.com/). The library itself is light, and so is their documentation. You don't need to read everything, use it mainly as a reference.
1. Update the above project we've been working on.
    1. Install `dotenv` package and implement environment variables for db connection information.
    1. Add search functionality via query parameters on the index route. For example, `GET /?search=sup` should return all usernames containing `sup`. DON'T implement this in JavaScript, search should be done in SQL.
    1. Add a new route `GET /delete` to delete all usernames from the db.
1. In our previous Mini Message Board project, we implemented ephemeral messages using an array i.e. the messages would reset when server restarted. We want data persistence. Go back to this project and implement it with a PostgreSQL db and `pg`.
   - Deploy a new db on a hosting service you choose, and obtain its connection information.
   - Create a `messages` table, populate it with data if you wish. This should be done via a script.

>    - Add necessary environment variables, create a pool, and implement required db functions.

Reviewer: Nit

```suggestion
   - Add the necessary environment variables, create a pool, and implement the required db functions.
```

</div>

### Knowledge check

The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.

> - [How to create a db and table via PostgreSQL shell?](#setting-up-the-db)
> - [What is node-postgres and how do you use it?](#using-node-postgres-in-express)
> - [When to use a client and a pool in pg?](#querying-with-pg)
> - [How would you integrate db query functions in your Express app?](#querying-with-pg)
> - [How to populate the db via a script?](#populate-the-db-via-a-script)

Reviewer: Nits
```suggestion
- [How do you create a db and table via PostgreSQL shell?](#setting-up-the-db)
- [What is node-postgres and how do you use it?](#using-node-postgres-in-express)
- [When should you use a client or a pool in pg?](#querying-with-pg)
- [How would you integrate db query functions in your Express app?](#querying-with-pg)
- [How would you populate the db via a script?](#populate-the-db-via-a-script)
```
